tests/test_0692_fsspec_writing.py

```diff
@@ -0,0 +1,107 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/uproot4/blob/main/LICENSE
+
+import pytest
+import uproot
+import uproot.source.fsspec
+
+import os
+import fsspec
+import numpy as np
+
+
+def test_fsspec_writing_no_integration(tmp_path):
+    uri = os.path.join(tmp_path, "some", "path", "file.root")
+    with fsspec.open(uri, mode="wb") as file_obj:
+        # write a simple root file
+        with uproot.recreate(file_obj) as f:
+            f["tree"] = {"x": np.array([1, 2, 3])}
+
+    with uproot.open(uri) as f:
+        assert f["tree"]["x"].array().tolist() == [1, 2, 3]
+
+
+@pytest.mark.parametrize("scheme", ["", "file://"])
+def test_fsspec_writing_local(tmp_path, scheme):
+    uri = scheme + os.path.join(tmp_path, "some", "path", "file.root")
+    with uproot.recreate(uri) as f:
+        f["tree"] = {"x": np.array([1, 2, 3])}
+
+    with uproot.open(uri) as f:
+        assert f["tree"]["x"].array().tolist() == [1, 2, 3]
+
+
+def test_fsspec_writing_http(server):
+    pytest.importorskip("aiohttp")
+
+    uri = f"{server}/file.root"
+
+    with pytest.raises(NotImplementedError):
+        # TODO: review this when fsspec supports writing to http
+        with uproot.recreate(uri) as f:
+            f["tree"] = {"x": np.array([1, 2, 3])}
+
+        with uproot.open(uri) as f:
+            assert f["tree"]["x"].array().tolist() == [1, 2, 3]
+
+
+@pytest.mark.parametrize(
+    "scheme",
+    [
+        "",
+        "file://",
```

Comment:
Can we add here also `simplecache::file://` per https://filesystem-spec.readthedocs.io/en/latest/features.html#remote-write-caching ? I expect this is the most common usage (e.g. `simplecache::s3://mybucket/myfile` or `simplecache::root://storage/path/to/file.root`)

Reply:
We talked about that in person yesterday. I don't think it's a good idea for Uproot to be smart and do the simplecache automatically. I can't think of a reason why someone would want to _not_ do simplecache, except perhaps to show how bad it is (as a line item in a performance table). But it would be a bad surprise if `root://` actually meant `samplecache::root://` (somebody gets surprised because they don't have 2 GB of disk space when they thought they were writing remotely), and if this automatically happens, there would be no way to get plain `root://` (for whatever reason they think they need it).

I think the best way to do this is for Uproot to naively do what is asked of it, but include the `simplecache::` prefix in all of the documentation. Most people will copy-paste.

Reply:
> Can we add here also simplecache::file://

I did not add it on the test because it doesn't currently work due to how path parsing is done. I will rewrite this to use fsspec for path parsing once fsspec is a hard requirement. (otherwise it works).

Reply:
Ok, so users are coached to use it, makes sense.
As for the test, this will need to wait until a future PR then?

Reply:
I will do some modifications so that it works with this PR because I think it's important that it works.

After trying to do the parsing with fsspec, I think it's still to hard to cover all the cases (due to the path:object syntax), I'm not sure it's even possible because you an have things such as: `simplecache::zip://insidezip://file///localfile.zip` (and add some other procotols with ports such as http to the mix...).

Reply:
Let's just not allow colons inside of the `object_path`. So it becomes a search for at most one colon, _not_ in a context that could be interpreted as a URI scheme, `::` namespace, or Windows drive letter.

Reply:
I don't think we are currently supporting colons inside the object path.

I just made some changes to update the split so it works with url chains but It's not going to support the full capabilities of fsspec. It's fine for this PR but I'll think of some solution.

Right now I like this solution (not 100% sure it's going to work though):

- Uproot can only be used to open files with `.root` extension.
- Object is specified after the root file in the urlpath (not necessarily at the end). For example: `            "simplecache::zip://uproot-issue121.root:Events/MET_pt::file:///tmp/pytest-of-runner/pytest-0/test_fsspec_zip0/uproot-issue121.root.zip"`. In this case the object is `Events/MET_pt`.
- Find the object by searching for a filename ending in `.root` followed by single `:`. (Will this be robust enough?).

Reply:
We can't assume that all ROOT files end with a `.root` extension, but we _could_ limit the applicability of the colon-parsing to just those files that do. (The colon-parsing is supposed to be a convenience. If you have a weird situation—a ROOT file not ending in `.root`—you don't deserve a convenience!)

The rule for what we do has to be simple enough to communicate and hopefully guess or intuit. Although the rules are complicated now, they're made that way to correspond to intuition. But how about the following?

1. If the string contains `.root:`, the colon of the _last_ `.root:` is used to split between URI-path and object-path.
2. Otherwise, the whole string is interpreted as a URI-path, every time.

This breaking change would have to go in when the minor version changed. Ideally, we would need to have warned users about this ahead of time, but I don't see a way to do that.

Reply:
The above would not remove the existing rules that `pathlib.Path` is entirely interpreted as a URI-path, every time, and the `{"uri-path": "object-path"}` syntax would still be recognized.

Reply:
I agree with those rules, they are simple enough and should be robust (we'll see if they can handle all cases, thankfully you cannot name a windows drive ".root"...).

- Find `.root` followed by single `:` (but not by `::`!)
- There shouldn't be any restriction to object paths except they cannot contain `:` (in that case use the dict notation).
- I think it makes more sense to apply the rule to the first `.root:` instead of the last given how fsspec chains urls, but I cannot think of any reasonable case where it would trigger more than once (a zip file having a .root extension instead of .zip...). In this case for instance matching the first appearance would make it work.

In any case I will make a separate PR for this into `main-fsspec` (which will be live next minor version).

Otherwise do you @nsmith- @jpivarski have any other comments regarding this PR? I'm not planning on further changes.

Reply:
I moved this to #1025.